### DISCUSSION TITLE
Self-hosting

### DISCUSSION THREAD

**Discussion by jvelo** [mfts/papermark#426] [General]
Hello there,

I gave a go to self-hosting on Vercel, I'm mostly there I think (haven't tried every feature yet)

To whom it may help, here's a couple of things I had to do:
- in `app/(static)/open-source-investors/page.tsx`, don't throw error on Airtable 401, but return an empty list
- in `lib/tinybird/pipes.ts`, change `get_total_average_page_duration` and `get_page_duration_per_view` pipes from `v5` to `v4` as the tb deploy step creates v4
- update the trigger.dev endpoint ID in `package.json` and `trigger.ts`
- in `lib/middleware/domain.ts`, change the redirect response to target domain
- in `middleware.ts`, add the target domain to the list of hosts for which to not return `DomainMiddleware`
- in `lib/resend.ts` change the "FROM" mails so that they can come from a verified domain I control
- in `pages/api/auth/[...nextauth].ts`, change the cookie domain for vercel deployments
- in `next.config.mjs`, tweak the patterns so that the target object store is whitelisted for images
- Add an `INTERNAL_API_KEY` env variable (I've set the value to a random secret)

Happy to give more details if needed, or to try and contribute to making it easier if there's interest!

**Comment by digitalethics** [comment]
@jvelo Thank you for the instructions. Do you know whether there is a possibility to replace Resend and Vercel with self-hostable open source alternatives?

**Comment by TSP-Dev** [comment]
i can't make it work. problems with tinybird and trigger. Can you explain what changes have to do?
ty
